### PR TITLE
HAI-2363/gis-automation-docker

### PR DESCRIPTION
**Pull request by petri-timlin**
# Description

Added docker image for gis automation.
Added docker_development section in config.yaml and usage for it.
Changed validate-deploy to use env variables in database connection.
Changed validate-deploy so that if tormays table is missing from database it will be created.

### Jira Issue: 
https://helsinkisolutionoffice.atlassian.net/browse/HAI-2363

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Other

# Instructions for testing
Instructions can be found from: scripts/gis-material-update/automation/README.md

### CODE REVIEW

File: scripts/gis-material-update/automation/automation.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#!/bin/sh
+
+for tormays_source in $*
```
Comment (corvidian) [inline scripts/gis-material-update/automation/automation.sh:3]: Running both `hsl` and `tram_lines` on the same run causes the script to download the HSL data twice?

File: scripts/gis-material-update/automation/automation.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+do
+    # Fetch data
+    echo "Fetching $tormays_source ..."
+    case $tormays_source in
+    hsl)
+        sh /haitaton-gis/fetch_all.sh $tormays_source hki
+        ;;
+    tram_infra)
+        sh /haitaton-gis/fetch_all.sh hki osm helsinki_osm_lines
```
Comment (corvidian) [inline scripts/gis-material-update/automation/automation.sh:12]: Looks like the `osm_vrt_clip.vrt` file is not copied to the docker image. 
```
❯ docker run --rm --network host --env-file .env haitaton-gis-automation tram_infra
Fetching tram_infra ...
Warning 6: Progress turned off as fast feature count is not available.
0...10...20...30...40...50...60...70...80...90...100 - done.
FAILURE:
Unable to open datasource `/haitaton-gis/osm_vrt_clip.vrt' with the following drivers.
  -> `FITS'
...
```

File: scripts/gis-material-update/automation/automation.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        sh /haitaton-gis/fetch_all.sh hki hsl
+        ;;
+    liikennevaylat)
+        sh /haitaton-gis/fetch_all.sh $tormays_source central_business_area ylre_katuosat
+        ;;
+    *)
+        sh /haitaton-gis/fetch_all.sh $tormays_source
+        ;;
+    esac
+
```
Comment (corvidian) [inline scripts/gis-material-update/automation/automation.sh:24]: Maybe the script should check the exit code of the previous step and only continue if it was 0?

File: scripts/gis-material-update/automation/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+WORKDIR /haitaton-gis-validate-deploy
+COPY ./config.yaml .
+COPY ./validate-deploy/validate_deploy_data.py .
+COPY ./validate-deploy/validate_deploy_data.sh .
+WORKDIR /haitaton-gis-validate-deploy/modules
+COPY ./validate-deploy/modules/ .
+WORKDIR /downloads
+WORKDIR /local_data
+WORKDIR /gis-output
+WORKDIR /haitaton-automation
```
Comment (corvidian) [inline scripts/gis-material-update/automation/Dockerfile:33]: If you specify the directory to copy to, you don't need to change the working directory all the time.
Empty directories can be created with `mkdir`.
```suggestion
COPY ./automation/automation.sh /haitaton-automation/
COPY ./config.yaml /haitaton-gis/
COPY ./config.yaml /haitaton-gis-validate-deploy/
COPY ./fetch/fetch_all.sh /haitaton-gis/
COPY ./fetch/fetch_data.sh /haitaton-gis/
COPY ./process/process_data.py /haitaton-gis/
COPY ./process/process_data.sh /haitaton-gis/
COPY ./process/modules/ /haitaton-gis/modules/
COPY ./validate-deploy/validate_deploy_data.py /haitaton-gis-validate-deploy/
COPY ./validate-deploy/validate_deploy_data.sh /haitaton-gis-validate-deploy/
COPY ./validate-deploy/modules/ /haitaton-gis-validate-deploy/modules/
RUN mkdir -p /downloads /gis-output /local_data
WORKDIR /haitaton-automation
```

File: scripts/gis-material-update/automation/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+RUN apt-get install -y software-properties-common wget gdal-bin libgdal-dev gettext-base python3-launchpadlib
+RUN add-apt-repository ppa:rmescandon/yq
+RUN apt install -y yq
+
+WORKDIR /opt
+COPY ./automation/requirements.txt /opt
+RUN python -m venv /opt/venv && /opt/venv/bin/pip install --upgrade pip && /opt/venv/bin/pip install -r requirements.txt
+WORKDIR /haitaton-automation
+COPY ./automation/requirements.txt .
```
Comment (corvidian) [inline scripts/gis-material-update/automation/Dockerfile:14]: This isn't used from `/haitaton-automation`.

File: scripts/gis-material-update/automation/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+pandas
+gtfs_kit
+pyyaml
+matplotlib
+sqlalchemy
+psycopg2-binary
+geoalchemy2
+pyjq
```
Comment (corvidian) [inline scripts/gis-material-update/automation/requirements.txt:8]: Are all of these requirements used? I can't find any references to some of them.

**Comment by umeetiusbaar** [comment]
Those Sonarcloud complains might also be good to check and fix if necessary. It provides also some good advice on how to fix them. And if they do not need fixing (e.g. running apt install -y), mark them as Safe.

File: scripts/gis-material-update/automation/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,35 @@
+FROM python:3.10
+
+ENV TORMAYS_DEPLOYMENT_PROFILE=docker_development
+RUN apt-get update
+
+RUN apt-get install -y software-properties-common wget gdal-bin libgdal-dev gettext-base python3-launchpadlib
+RUN add-apt-repository ppa:rmescandon/yq
+RUN apt install -y yq
```
Comment (corvidian) [inline scripts/gis-material-update/automation/Dockerfile:8]: The best practices say that:
1. Don't install recommended packages.
2. Clean cache after installing, to remove unnecessary files from the built image.
3. Add the dependencies one per row in alphabetical order. This makes it easier to track changes to them.
4. Don't mix apt-get and apt, it's just confusing.
```suggestion
RUN apt-get update  \
    && apt-get install --no-install-recommends -y \
    gdal-bin \
    gettext-base \
    libgdal-dev \
    python3-launchpadlib  \
    software-properties-common \
    wget \
    && apt-get clean
RUN add-apt-repository ppa:rmescandon/yq
RUN apt-get install --no-install-recommends -y yq && apt-get clean
```

File: scripts/gis-material-update/automation/requirements.txt (hunk 6 above)
Comment (corvidian) [reply scripts/gis-material-update/automation/requirements.txt:8]: This set of requirements was enough to run all the current targets:
```
pandas
gtfs_kit
pyyaml
sqlalchemy
psycopg2-binary
geoalchemy2
parse
```

File: scripts/gis-material-update/automation/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+## Run image
+Give following:
+```sh
+docker run --rm --network host --env-file haitaton.env haitaton-gis-automation <source_1> ... <source_N>
+```
+
+Where `<source>` is currently one of:
+
+- `hsl` - HSL bus schedules
```
[the remark this answers is not in the document]
Comment (petri-timlin) [reply scripts/gis-material-update/automation/README.md:41]: Changed to HSL bus routes and added tram_lines and tram_infra

File: scripts/gis-material-update/automation/automation.sh (hunk 1 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/automation.sh:3]: Yes it does. There is no easy way to check in this level (bash script) that is the data already downloaded or not.

Comment (corvidian) [reply scripts/gis-material-update/automation/automation.sh:3]: You could have the check in `fetch_data.sh`. Maybe with an option to override the file anyway.

Another approach would be to run all the fetch phases at once, then all the process and deploy phases. Maybe something like (untested):
```shell
# Collect the sources needed for all targets:
sources=""
for tormays_source in $*
do
    case $tormays_source in
    hsl)
        sources="${sources} hsl hki"
        ;;
    tram_infra)
        sources="${sources} hki osm helsinki_osm_lines"
        ;;
    tram_lines)
        sources="${sources} hki hsl"
        ;;
    liikennevaylat)
        sources="${sources} liikennevaylat central_business_area ylre_katuosat"
        ;;
    *)
        sources="${sources} $tormays_source"
        ;;
    esac
done
# Deduplicate sources
sources=$echo "$sources" | xargs -n1 | sort -u | xargs)
# Fetch the sources
sh /haitaton-gis/fetch_all.sh ${sources}
```

Comment (petri-timlin) [reply scripts/gis-material-update/automation/automation.sh:3]: sources=$(echo "$sources" | xargs -n1 | sort -u | xargs)

File: scripts/gis-material-update/automation/Dockerfile (hunk 7 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/Dockerfile:8]: Change done.

File: scripts/gis-material-update/automation/Dockerfile (hunk 4 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/Dockerfile:33]: Change done.

File: scripts/gis-material-update/automation/requirements.txt (hunk 6 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/requirements.txt:8]: Change done.

File: scripts/gis-material-update/automation/automation.sh (hunk 2 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/automation.sh:12]: Added osm_vrt_clip.vrt file to docker

File: scripts/gis-material-update/automation/automation.sh (hunk 1 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/automation.sh:3]: Changed

File: scripts/gis-material-update/automation/automation.sh (hunk 3 above)
Comment (petri-timlin) [reply scripts/gis-material-update/automation/automation.sh:24]: Added error handling

**Review by corvidian — APPROVED** [review]

**Pull request merged**